### PR TITLE
EL-4218 - Select custom heading options

### PR DESCRIPTION
**Pull request by cmallon367**
#### Checklist

* [x] The contents of this PR are mine to submit. No third party code or other material is being committed.
* [x] New third party dependencies (if any) are linked via `package.json`. Third party dependencies are licensed under one of the following: Apache, MIT, BSD, Mozilla Public License, Eclipse Public License, or Oracle Binary Code License.
* [x] The code in this PR does not contain export-restricted encryption algorithms.

#### Ticket / Issue

https://portal.digitalsafe.net/browse/EL-4218

#### Description of Proposed Changes

Added new template bindings
Updated the documentation to include new template bindings
Added additional test coverage and screenshot to show the custom headings
#### Documentation CI URL

https://pages.github.houston.softwaregrp.net/sepg-docs-qa/UXAspects_CI_UXAspects_EL-4218-select-custom-headings

#### Downstream Build(s)

* [caf/ux-aspects-micro-focus](https://jenkins.swinfra.net/job/SEPG/job/caf/job/caf~ux-aspects-micro-focus~EL-4218-select-custom-headings~CI/)

### CODE REVIEW

File: src/components/select/select.component.ts [hunk 1]
```diff
@@ -134,6 +134,10 @@ export class SelectComponent<T> implements OnInit, OnChanges, OnDestroy, Control
      */
     @Input() tagTemplate: TemplateRef<TagTemplateContext>;
 
+    @Input() headingTemplate: TemplateRef<any>;
```
Comment (donna-marijne) [inline src/components/select/select.component.ts:137]: Don't use `any`.

File: src/components/typeahead/typeahead.component.html [hunk 2]
```diff
@@ -74,4 +84,8 @@
 
 <ng-template #defaultNoOptionsTemplate>
     <span class="ux-typeahead-no-options">No results</span>
-</ng-template>
\ No newline at end of file
+</ng-template>
+
+<ng-template #defaultRecentOptionsHeadingTemplate></ng-template>
+
+<ng-template #defaultHeadingTemplate></ng-template>
```
Comment (donna-marijne) [inline src/components/typeahead/typeahead.component.html:91]: These are empty, so remove them.

File: src/components/typeahead/typeahead.component.less [hunk 3]
```diff
@@ -85,4 +85,11 @@ ux-typeahead {
         line-height: 24px;
         font-style: italic;
     }
+
+    .ux-typeahead-recent-options-heading {
```
Comment (donna-marijne) [inline src/components/typeahead/typeahead.component.less:89]: This isn't used.

File: docs/app/pages/components/components-sections/select/select/select.component.html [hunk 4]
```diff
@@ -275,6 +275,16 @@
     <tr uxd-api-property name="autoCloseDropdown" type="boolean" defaultValue="true">
         Determine if the dropdown panel should close on external click.
     </tr>
+    <tr uxd-api-property name="headingTemplate" type="TemplateRef">
+        <p>A template which will display custom headings for normal options.</p>
+        <p>
+            The following code snippet shows how to use a custom template to display an icon alongside the tag text:
```
Comment (donna-marijne) [inline docs/app/pages/components/components-sections/select/select/select.component.html:281]: This description is incorrect.

File: docs/app/pages/components/components-sections/select/select/select.component.html [hunk 5]
```diff
@@ -275,6 +275,16 @@
     <tr uxd-api-property name="autoCloseDropdown" type="boolean" defaultValue="true">
         Determine if the dropdown panel should close on external click.
     </tr>
+    <tr uxd-api-property name="headingTemplate" type="TemplateRef">
```
Comment (donna-marijne) [inline docs/app/pages/components/components-sections/select/select/select.component.html:278]:
```suggestion
    <tr uxd-api-property name="optionsHeadingTemplate" type="TemplateRef">
```
`headingTemplate` is too generic a name for a component like this.

File: e2e/tests/components/select/custom-heading/select-custom-heading.e2e-spec.ts [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+    it('should display custom headings for recent and normal options', async () => {
+        await page.clickOnCheckbox(page.checkboxRecentOptions);
+        await page.clickOnDropdown(false);
+
+        await page.clickOnCountry(false, 1);
+        await page.checkRecentOptions(false, ['United Kingdom']);
+
+        expect(await imageCompare('select-custom-heading')).toEqual(0);
+    });
```
Comment (donna-marijne) [inline e2e/tests/components/select/custom-heading/select-custom-heading.e2e-spec.ts:21]: You should add coverage for when there are no recent options selected.

**Review by donna-marijne — CHANGES_REQUESTED** [review]
It doesn't look like the options heading shows when there aren't any recent options?

https://plnkr.co/edit/scgH5E6aXtnZl57N

Please take time to look over your changes and make sure that they work before submitting for review. Seeing a lot of mistakes from copy/paste in here.

File: e2e/pages/app/select/custom-heading/select-custom-heading.testpage.component.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+export class SelectCustomHeadingTestPageComponent implements OnInit {
+
+    // ux-select configuration properties
+    options: string[] | Function;
+    dropdownOpen: boolean;
+    placeholder = 'Select a country';
+    recentOptions: ReadonlyArray<string> = null;
+
+    // Customize settings
+    dataSet = new BehaviorSubject<string>('strings');
```
Comment (donna-marijne) [inline e2e/pages/app/select/custom-heading/select-custom-heading.testpage.component.ts:18]: Can you remove this unused code?

File: e2e/tests/components/select/custom-heading/select-custom-heading.po.spec.ts [hunk 8]
```diff
@@ -0,0 +1,70 @@
+import { browser, by, element, ElementFinder } from 'protractor';
+
+export class SelectCustomHeadingPage {
```
Comment (donna-marijne) [inline e2e/tests/components/select/custom-heading/select-custom-heading.po.spec.ts:3]: Can you remove the unused code in this class? Additionally, all functions should have a return type.

File: e2e/tests/components/select/custom-heading/select-custom-heading.po.spec.ts [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+    dropdown = element(by.id('dropdown'));
+    nodes = element(by.id('Nodes'));
```
Comment (donna-marijne) [inline e2e/tests/components/select/custom-heading/select-custom-heading.po.spec.ts:6]:
```suggestion
    optionsHeading = element(by.id('Nodes'));
```

File: e2e/pages/app/select/custom-heading/select-custom-heading.testpage.component.html [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+        <div>
+            <ux-checkbox id="checkbox1" (valueChange)="recentOptions = $event ? [] : undefined">
+                recentOptions
+            </ux-checkbox>
+        </div>
+    </div>
+</div>
+
+<ng-template #demoHeader>
+    <div id="Nodes">Nodes</div>
```
Comment (donna-marijne) [inline e2e/pages/app/select/custom-heading/select-custom-heading.testpage.component.html:22]:
```suggestion
    <div id="options-header">Nodes</div>
```
Try to use meaningful IDs/variable names.

File: e2e/pages/app/select/custom-heading/select-custom-heading.testpage.component.html [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+</ng-template>
+
+<ng-template #demoRecentHeaderOptions>
+    <div id="recentNodes">Recent Nodes</div>
```
Comment (donna-marijne) [inline e2e/pages/app/select/custom-heading/select-custom-heading.testpage.component.html:26]:
```suggestion
    <div id="recent-options-header">Recent Nodes</div>
```

File: src/components/typeahead/typeahead.component.less [hunk 12]
```diff
@@ -66,6 +66,12 @@ ux-typeahead {
                 }
             }
         }
+
+        .ux-typeahead-recent-divider {
```
Comment (donna-marijne) [inline src/components/typeahead/typeahead.component.less:70]: You haven't used this class.

**Review by donna-marijne — CHANGES_REQUESTED** [review]
Looks like the divider is broken:
image

https://plnkr.co/edit/yhiijelTSdZSqxws

Check the suggestion from the previous review.

The new inputs for typeahead have not been documented:
https://pages.github.houston.softwaregrp.net/sepg-docs-qa/UXAspects_CI_UXAspects_EL-4218-select-custom-headings/#/components/input-controls#typeahead

**Review by JonathanToddMF — DISMISSED** [review]
Thanks Christina looks good!

**Review by donna-marijne — APPROVED** [review]
This should be good for QA now.

**Comment by JonathanToddMF** [comment]
QA Complete - Tested on Chrome, Firefox, Edge and IE11

Following plunker - https://plnkr.co/edit/DAPfEATGqibo6h0P shows the recent options heading and options heading.
Note: Styling of these headings would be up to the implementer. I have added some padding to the headings.

Awaiting acceptance @Roland-Penny

**Review by Roland-Penny — APPROVED** [review]
Accepted

**Pull request merged**